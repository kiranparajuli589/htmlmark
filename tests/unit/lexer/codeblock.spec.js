import Lexer from "../../../lib/lexer/index.js"


describe("codeblock", () => {
	it("should parse the codeblock", () => {
		const lines = [
			"```js",
			"const a = 1",
			"```"
		]
		const lexer = new Lexer(lines)
		const tokens = lexer.run()
		expect(tokens).toMatchSnapshot()
	})
	it("should parse the codeblock without the language", () => {
		const lines = [
			"```",
			"const a = 1",
			"```"
		]
		const lexer = new Lexer(lines)
		const tokens = lexer.run()
		expect(tokens).toMatchSnapshot()
	})
	it("should parse the codeblock with multiple lines", () => {
		const lines = [
			"```js",
			"const a = 1",
			"const b = 2",
			"```"
		]
		const lexer = new Lexer(lines)
		const tokens = lexer.run()
		expect(tokens).toMatchSnapshot()
	})
	it("should parse the codeblock with multiple lines and a newline", () => {
		const lines = [
			"```js",
			"const a = 1",
			"  const b = 2",
			"",
			"const c = 3",
			"```"
		]
		const lexer = new Lexer(lines)
		const tokens = lexer.run()
		expect(tokens).toMatchSnapshot()
	})
	it("should parse multiple consecutive codeblocks", () => {
		const lines = [
			"```js",
			"const a = 1",
			"```",
			"```js",
			"const b = 2",
			"```"
		]
		const lexer = new Lexer(lines)
		const tokens = lexer.run()
		expect(tokens).toMatchSnapshot()
	})
	it("should cope with multiple items", () => {
		const lines = [
			"```js",
			"const a = 1",
			"```",
			"some people are funny"
		]
		const lexer = new Lexer(lines)
		const tokens = lexer.run()
		expect(tokens).toMatchSnapshot()
	})
	it("should set incomplete codeblock as a code block", () => {
		const lines = [
			"some line",
			"    ```js",
			"    const a = 1",
			"    const b = 2",
			"here again" // broken indentation
		]
		const lexer = new Lexer(lines)
		const tokens = lexer.run()
		expect(tokens).toMatchSnapshot()
	})
	it("should detect mis-indented closing codeblock", () => {
		const lines = [
			"```js",
			"const a = 1",
			"const b = 2",
			"const c = 3",
			"    ```"
		]
		const lexer = new Lexer(lines)
		const tokens = lexer.run()
		expect(tokens).toMatchSnapshot()
	})
	it("should detect acceptable indent for closing codeblock", () => {
		const lines = [
			"```js",
			"const a = 1",
			"const b = 2",
			"const c = 3",
			"   ```"
		]
		const lexer = new Lexer(lines)
		const tokens = lexer.run()
		expect(tokens).toMatchSnapshot()
	})
	it("should not be broken with acceptable wrong indent", () => {
		const lines = [
			"   ```",
			"abcd",
			"   ```"
		]
		const lexer = new Lexer(lines)
		const tokens = lexer.run()
		expect(tokens).toMatchSnapshot()
	})
	it("should cope with empty body", () => {
		const lines = [
			"```js",
			"```"
		]
		const lexer = new Lexer(lines)
		const tokens = lexer.run()
		expect(tokens).toMatchSnapshot()
	})
	it("should be set just with indentation bigger or equals to 4 at top", () => {
		const lines = [
			"    one",
			"          two",
			"    three",
			"four"
		]
		const lexer = new Lexer(lines)
		const tokens = lexer.run()
		expect(tokens).toMatchSnapshot()
	})
	it("should be set just with indentation bigger or equals to 4 below newline", () => {
		const lines = [
			"zero one two",
			"",
			"    one",
			"          two",
			"    three",
			"four"
		]
		const lexer = new Lexer(lines)
		const tokens = lexer.run()
		expect(tokens).toMatchSnapshot()
	})
	it("should be set just with indentation bigger or equals to 4 but consecutive with other tokens", () => {
		const lines = [
			"zero one two",
			"    one",
			"          two",
			"    three",
			"four"
		]
		const lexer = new Lexer(lines)
		const tokens = lexer.run()
		expect(tokens).toMatchSnapshot()
	})
	it("should be work just fine within a list item", () => {
		const testString = `
4.  What is the exact rule for determining when list items get
\twrapped in \`<p>\` tags?  Can a list be partially "loose" and partially
\t"tight"?  What should we do with a list like this?

\t\t\`\`\` markdown
\t\t1. one

\t\t2. two
\t\t3. three
\t\t\`\`\`
`
		const lines = testString.split("\n")
		const lexer = new Lexer(lines)
		const tokens = lexer.run()
		expect(tokens).toMatchSnapshot()
	})
	it("should remove the link refs inside codeblock but do not change the usage", () => {
		const lines = `\`\`\` markdown
[foo]: /url1
[foo]: /url2

[foo][]
\`\`\``
		const lexer = new Lexer(lines.split("\n"))
		const tokens = lexer.run()
		expect(tokens).toMatchSnapshot()
	})
})
