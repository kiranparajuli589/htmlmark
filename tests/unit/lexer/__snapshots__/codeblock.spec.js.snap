// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codeblock should be set just with indentation bigger or equals to 4 at top 1`] = `
Array [
  Object {
    "indent": 4,
    "language": null,
    "raw": "    one
          two
    three",
    "type": "code-block",
    "value": "one
      two
three",
  },
  Object {
    "indent": 0,
    "raw": "four",
    "tokens": Array [
      Object {
        "raw": "four",
        "type": "text",
        "value": "four",
      },
    ],
    "type": "paragraph",
    "value": "four",
  },
]
`;

exports[`codeblock should be set just with indentation bigger or equals to 4 below newline 1`] = `
Array [
  Object {
    "indent": 0,
    "raw": "zero one two",
    "tokens": Array [
      Object {
        "raw": "zero one two",
        "type": "text",
        "value": "zero one two",
      },
    ],
    "type": "paragraph",
    "value": "zero one two",
  },
  Object {
    "type": "new-line",
  },
  Object {
    "indent": 4,
    "language": null,
    "raw": "    one
          two
    three",
    "type": "code-block",
    "value": "one
      two
three",
  },
  Object {
    "indent": 0,
    "raw": "four",
    "tokens": Array [
      Object {
        "raw": "four",
        "type": "text",
        "value": "four",
      },
    ],
    "type": "paragraph",
    "value": "four",
  },
]
`;

exports[`codeblock should be set just with indentation bigger or equals to 4 but consecutive with other tokens 1`] = `
Array [
  Object {
    "indent": 0,
    "raw": "zero one two
    one
          two
    three
four",
    "tokens": Array [
      Object {
        "raw": "zero one two     one           two     three four",
        "type": "text",
        "value": "zero one two     one           two     three four",
      },
    ],
    "type": "paragraph",
    "value": "zero one two     one           two     three four",
  },
]
`;

exports[`codeblock should cope with empty body 1`] = `
Array [
  Object {
    "indent": 0,
    "language": "js",
    "raw": "\`\`\`js
\`\`\`",
    "type": "code-block",
    "value": "",
  },
]
`;

exports[`codeblock should cope with multiple items 1`] = `
Array [
  Object {
    "indent": 0,
    "language": "js",
    "raw": "\`\`\`js
const a = 1
\`\`\`",
    "type": "code-block",
    "value": "const a = 1",
  },
  Object {
    "indent": 0,
    "raw": "some people are funny",
    "tokens": Array [
      Object {
        "raw": "some people are funny",
        "type": "text",
        "value": "some people are funny",
      },
    ],
    "type": "paragraph",
    "value": "some people are funny",
  },
]
`;

exports[`codeblock should detect acceptable indent for closing codeblock 1`] = `
Array [
  Object {
    "indent": 0,
    "language": "js",
    "raw": "\`\`\`js
const a = 1
const b = 2
const c = 3
   \`\`\`",
    "type": "code-block",
    "value": "const a = 1
const b = 2
const c = 3
   \`\`\`",
  },
]
`;

exports[`codeblock should detect mis-indented closing codeblock 1`] = `
Array [
  Object {
    "indent": 0,
    "language": "js",
    "raw": "\`\`\`js
const a = 1
const b = 2
const c = 3
    \`\`\`",
    "type": "code-block",
    "value": "const a = 1
const b = 2
const c = 3
    \`\`\`",
  },
]
`;

exports[`codeblock should not be broken with acceptable wrong indent 1`] = `
Array [
  Object {
    "indent": 2,
    "language": null,
    "raw": "   \`\`\`",
    "type": "code-block",
    "value": "",
  },
  Object {
    "indent": 0,
    "raw": "abcd",
    "tokens": Array [
      Object {
        "raw": "abcd",
        "type": "text",
        "value": "abcd",
      },
    ],
    "type": "paragraph",
    "value": "abcd",
  },
  Object {
    "indent": 2,
    "language": null,
    "raw": "   \`\`\`",
    "type": "code-block",
    "value": "",
  },
]
`;

exports[`codeblock should parse multiple consecutive codeblocks 1`] = `
Array [
  Object {
    "indent": 0,
    "language": "js",
    "raw": "\`\`\`js
const a = 1
\`\`\`",
    "type": "code-block",
    "value": "const a = 1",
  },
  Object {
    "indent": 0,
    "language": "js",
    "raw": "\`\`\`js
const b = 2
\`\`\`",
    "type": "code-block",
    "value": "const b = 2",
  },
]
`;

exports[`codeblock should parse the codeblock 1`] = `
Array [
  Object {
    "indent": 0,
    "language": "js",
    "raw": "\`\`\`js
const a = 1
\`\`\`",
    "type": "code-block",
    "value": "const a = 1",
  },
]
`;

exports[`codeblock should parse the codeblock with multiple lines 1`] = `
Array [
  Object {
    "indent": 0,
    "language": "js",
    "raw": "\`\`\`js
const a = 1
const b = 2
\`\`\`",
    "type": "code-block",
    "value": "const a = 1
const b = 2",
  },
]
`;

exports[`codeblock should parse the codeblock with multiple lines and a newline 1`] = `
Array [
  Object {
    "indent": 0,
    "language": "js",
    "raw": "\`\`\`js
const a = 1
  const b = 2

const c = 3
\`\`\`",
    "type": "code-block",
    "value": "const a = 1
  const b = 2

const c = 3",
  },
]
`;

exports[`codeblock should parse the codeblock without the language 1`] = `
Array [
  Object {
    "indent": 0,
    "language": null,
    "raw": "\`\`\`
const a = 1
\`\`\`",
    "type": "code-block",
    "value": "const a = 1",
  },
]
`;

exports[`codeblock should set incomplete codeblock as a code block 1`] = `
Array [
  Object {
    "indent": 0,
    "raw": "some line",
    "tokens": Array [
      Object {
        "raw": "some line",
        "type": "text",
        "value": "some line",
      },
    ],
    "type": "paragraph",
    "value": "some line",
  },
  Object {
    "indent": 4,
    "language": "js",
    "raw": "    \`\`\`js
    const a = 1
    const b = 2",
    "type": "code-block",
    "value": "const a = 1
const b = 2",
  },
  Object {
    "indent": 0,
    "raw": "here again",
    "tokens": Array [
      Object {
        "raw": "here again",
        "type": "text",
        "value": "here again",
      },
    ],
    "type": "paragraph",
    "value": "here again",
  },
]
`;
